platform-sdk/docs/proposals/metric-labels/README.md

```diff
@@ -0,0 +1,343 @@
+- Make the metrics module independent of the platform module
+- Provide a public API to define labels for metrics
+- Add labels to the snapshot API
+- The output of the snapshot API should contain the labels
+- The output of the snapshot API must be compatible with the current output (no breaking changes)
+- Create better and more dynamic dashboards in Grafana by using labels
+
+## Non-Goals
+
+Even with the new concept the metrics api should be only used in our (hashgraph) code.
+The metrics module should not be used by the end user.
+Here the preferred solution is still that we forward our metrics to a monitoring system that the custom application code uses.
+
+## Design & Architecture
+
+This chapter describes the changes that are needed to add labels to the metrics module.
+
+### About labels
+
+Before we start with the implementation we need to define what labels are.
+Labels are key-value pairs that are used to define, filter and group metrics.
+Labels are used by monitoring systems like Grafana to create more dynamic dashboards.
+By adding labels to the metrics module a single metric is now uniquely defined by its name and the labels.
+For example, metrics that counts the number of requests can have a label called `method` that defines if the request was a `GET` or `POST` request.
+That metrics are defined by the name "api_http_requests_total" and the label `method`.
+By doing so the monitoring system can create a dashboard that shows the number of requests per method.
+In the given example we would have 2 metric instances: one for `GET` and one for `POST`:
+
+- Metric 1: `api_http_requests_total{method="GET"}`
+- Metric 2: `api_http_requests_total{method="POST"}`
+
+While today the name of a metric is the unique identifier of a metric, the unique identifier will become the name and the labels once labels are added.
+By doing so both metrics can use the same name but are still unique as long as the labels are different.
+We will store measurements individually for each metric.
+If you later want to know the combined measurement of all metrics with the name `api_http_requests_total` you can use the prometheus query language to sum up the values of all metrics with that name.
+See https://prometheus.io/docs/prometheus/latest/querying/basics/#time-series-selectors for more information.
+
+In prometheus a concrete metric is defined like this (see https://prometheus.io/docs/concepts/data_model/#notation):
+
+```
+api_http_requests_total{method="POST", handler="/messages"}
+```
+
+Here a metric is defined by a name (`api_http_requests_total`) and a set of labels (`method="POST"` and `handler="/messages"`).
+In prometheus that defines the unique identifier of a metric.
+
+When using labels to create dashboards in Grafana prometheus provides a rich query language to filter and group metrics: https://prometheus.io/docs/prometheus/latest/querying/examples/
+
+We will use the same practices and definition as in prometheus regarding labels.
```

Comment:
Might be good to either link to those practice or call out some important ones to avoid.

Reply:
@poulok can I add the links without restart voting?

- https://prometheus.io/docs/practices/naming/#labels
 - https://prometheus.io/docs/practices/instrumentation/#use-labels

Reply:
done